proto/cacheclient.proto

```diff
@@ -402,16 +402,110 @@ message _ListRemoveRequest {
 
 message _ListRemoveResponse {}
 
+/*
+  # ListFetch and ListRetain documentation and examples
+
+  # start and end are zero-based indexes, both are optional
+
+  # start is inclusive
+  # If start is not set, then the default is Unbounded, which is treated as 0
+  # Unbounded == 0,
+  # this field has a explicit presence for future compatibility to support reverse order and controlling our stride
+
+  # end is exclusive
+  # If end is not set, then the default is Unbounded which becomes (list.len)
+
+  # start and end can be negative numbers, negative indicates offsets starting at the end of the list.
+  # For example, -1 is the last element of the list, -2 the penultimate, etc
+  #               [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # +start/+end    0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+  # -start/-end  -10,-9,-8,-7,-6,-5,-4,-3,-2,-1
+
+  list_ = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # Using the slice notation:
+  # start:end
+
+  # start=0, end=Unbounded
+  list_[0:] == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # start=5, end=Unbounded
+  list_[5:] == [5, 6, 7, 8, 9]
+
+  # start=5, end=0
+  list_[5:0] == []
+
+  # start=-5, end=Unbounded
+  list_[-5:] == [5, 6, 7, 8, 9]
+
+  # start=-5, end=0
+  list_[-5:0] == []
+
+  # start=0, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=Unbounded, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=0, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=Unbounded, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=2, end=6
+  list_[2:6] == [2, 3, 4, 5]
+
+  # start=2, end=-6
+  list_[2:-6] == [2, 3]
+
+  # start=-7, end=-2
+  list_[-7:-2] == [3, 4, 5, 6, 7]
+
+  // if using ltrim, this will remove all elements in the list
+  # start=-2, end=-6
+  list_[2:-6] == []
+
+  # empty list
+  list_[-1:1] == []
+ */
```

Comment:
I think this is wrong / only true on empty lists and lists greater than length 1. Consider a list containing `[a]`. What is `[-1:1]`? My brain says it would be `[a]` and Python agrees it's `['a']`.
```
$ python3 -c 'print(["a"][-1:1])'
['a']
```

I assert the correct empty range definition is actually `[0:0]`. That directly implies that the exclusivity at the end of the range needs to take precedence over the inclusivity at the start of the range.

Can you provide a set of expectations for lists of length 1 with respect to what special range cases mean? I'm not sure if lists of length 2 will also have special considerations.

Reply:
Agreed. The other convention to document is that we implicitly clip out of range end indexes to `list.len`. Ie in this case there is no list element with index `1`, so end=1 implies `end=Unbounded`

Reply:
Oh good catch.

Looks like list of length 2 aren't affected by the same special cases